h2o-py/h2o/expr.py

```diff
@@ -121,24 +120,44 @@ def _eval_driver(self, top):
     # the timezone change, so cannot be lazy.
     #
     def _get_ast_str(self, top):
-        if not self._cache.is_empty():  # Data already computed and cached; could a "false-like" cached value
-            return str(self._cache._data) if self._cache.is_scalar() else self._cache._id
-        if self._cache._id is not None:
-            return self._cache._id  # Data already computed under ID, but not cached
-        # assert isinstance(self._children,tuple)
-        exec_str = "({} {})".format(self._op, " ".join([ExprNode._arg_to_expr(ast) for ast in self._children]))
-        gc_ref_cnt = len(gc.get_referrers(self))
-        if top or gc_ref_cnt >= ExprNode.MAGIC_REF_COUNT:
-            self._cache._id = _py_tmp_key(append=h2o.connection().session_id)
-            exec_str = "(tmp= {} {})".format(self._cache._id, exec_str)
-        return exec_str
+        stack = collections.deque()
+        # Append self and evaluate
+        stack.append(self)
+        r = collections.deque()
+        while len(stack) > 0:
+            head = stack.pop()
+            s = None
+            if isinstance(head, ExprNode):
+                if not head._cache.is_empty():  # Data already computed and cached; could a "false-like" cached value
+                    s = str(head._cache._data) if head._cache.is_scalar() else head._cache._id
+                elif head._cache._id is not None:
+                    s = head._cache._id  # Data already computed under ID, but not cached
+                else:
+                    s = "({}".format(head._op)
+                    gc_ref_cnt = len(gc.get_referrers(head))
+                    if (top and head == self) or gc_ref_cnt >= ExprNode.MAGIC_REF_COUNT:
```

Comment:
Does this step mean that if `top` is true ( -> cache it) and we are on the top-level frame, do cache?